File: flutter_app/lib/domain/progress.dart
import 'models.dart';

bool isHoleRequiredComplete(HoleScore hole, RoundConfig config) {
  void config;
  return hole.strokes != null;
}

int findFirstIncompleteHole(RoundRecord record) {
  for (final hole in record.holes) {
    if (!isHoleRequiredComplete(hole, record.round.configSnapshot)) {
      return hole.holeNo;
    }
  }
  return record.round.holesCount;
}

bool isRoundReadyToComplete(RoundRecord record) {
  return record.holes
      .every((hole) => isHoleRequiredComplete(hole, record.round.configSnapshot));
}

File: lib/domain/progress.ts
import type { HoleScore, RoundConfig, RoundRecord } from "@/lib/domain/types";

export function isHoleRequiredComplete(hole: HoleScore, config: RoundConfig) {
  void config;
  return hole.strokes != null;
}

export function findFirstIncompleteHole(record: RoundRecord) {
  return (
    record.holes.find((hole) => !isHoleRequiredComplete(hole, record.round.configSnapshot))
      ?.holeNo ?? record.round.holesCount
  );
}

export function isRoundReadyToComplete(record: RoundRecord) {
  return record.holes.every((hole) => isHoleRequiredComplete(hole, record.round.configSnapshot));
}
